File: grind75/Dec21/110.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        
        def depth(root):
            if not root:
                return 0
            return 1 + max(depth(root.left), depth(root.right))
        
        if not root:
            return True

        if abs(depth(root.left) - depth(root.right)) > 1:
            return False

        return self.isBalanced(root.left) and self.isBalanced(root.right)      

File: grind75/Dec21/110.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isBalanced(root *TreeNode) bool {
    var dfs func(*TreeNode) int
    dfs = func(root *TreeNode) int {
        if root == nil { return 0 }
        l, r := dfs(root.Left), dfs(root.Right)
        if l == -1 || r == -1 { return -1 }
        if math.Abs(float64(l-r)) > 1 { return -1 }
        return max(l,r) + 1
    }
    ret := dfs(root)
    return ret != -1
}
